src/main/haskell/kore/src/Kore/Predicate/Predicate.hs

```diff
@@ -378,3 +382,49 @@ allVariables
     => Predicate level variable
     -> Set (variable level)
 allVariables = pureAllVariables . unwrapPredicate
+
+{- | Extract the set of free variables from a @Predicate@.
+-}
+freeVariables
+    :: ( MetaOrObject level
+       , Show (variable Object)
+       , Show (variable Meta)
+       , Ord (variable Object)
+       , Ord (variable Meta))
+    => Predicate level variable
+    -> Set (variable level)
+freeVariables = pureFreeVariables Proxy . unwrapPredicate
+
+{- | 'substitutionToPredicate' transforms a substitution in a predicate.
+
+    An empty substitution list returns a true predicate. A non-empty
+    substitution returns a conjunction of variable/substitution equalities.
+
+-}
+substitutionToPredicate
+    ::  ( MetaOrObject level
+        , Given (SymbolOrAliasSorts level)
+        , SortedVariable variable
+        , Eq (variable level)
+        , Show (variable level))
+    => [(variable level, PureMLPattern level variable)]
+    -> Predicate level variable
+substitutionToPredicate =
+    foldl'
```

Comment:
Now, that I'm re-reading this code, it may have been cleaner as a map + makeMultipleAndPredicate. This is not a request for changes, I know it was just moved from a different file.